source/common/upstream/upstream_connection_pool_impl.cc

```diff
@@ -0,0 +1,173 @@
+#include "common/upstream/upstream_connection_pool_impl.h"
+
+#include "common/common/lock_guard.h"
+
+namespace Envoy {
+namespace Upstream {
+
+UpstreamConnectionPoolImpl::UpstreamConnectionPoolImpl(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {}
+
+UpstreamConnectionPoolImpl::~UpstreamConnectionPoolImpl() {
+  for (auto& pool : connection_pools_) {
+    auto& clients = pool.second.clients;
+    while (!clients.empty()) {
+      clients.front()->client_connection_->close(Network::ConnectionCloseType::NoFlush);
+    }
+  }
+
+  // Make sure all clients are destroyed before we are destroyed.
+  dispatcher_.clearDeferredDeleteList();
+}
+
+void UpstreamConnectionPoolImpl::OfferConnection(HostConstSharedPtr host,
+                                                 UpstreamConnectionEssence connection_essence) {
+  auto shared_connection_essence =
+      std::make_shared<UpstreamConnectionEssence>(std::move(connection_essence));
+  dispatcher_.post([this, host, shared_connection_essence]() {
```

Comment:
What happens in cases where the dispatcher's event loop is already stopped or dispatcher is otherwise unable to accept posted closures?  Seems like we have some sporadic shutdown related crashes in //test/integration:h1_capture_fuzz_test with this change.

avd@kallisti $ bazel test --runs_per_test=1000 //test/integration:h1_capture_fuzz_test
//test/integration:h1_capture_fuzz_test                                  FAILED in 6 out of 1000 in 3.3s
  Stats over 1000 runs: max = 3.3s, min = 0.7s, avg = 3.0s, dev = 0.2s
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_196_of_1000/test.log
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_39_of_1000/test.log
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_189_of_1000/test.log
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_398_of_1000/test.log
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_369_of_1000/test.log
  /home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/integration/h1_capture_fuzz_test/run_547_of_1000/test.log

[2019-11-06 18:58:29.774][38][info][upstream] [source/server/lds_api.cc:63] lds: add/update listener 'http'
[2019-11-06 18:58:29.774][38][info][config] [source/server/listener_manager_impl.cc:548] all dependencies initialized. starting workers
[2019-11-06 18:58:29.774][38][info][main] [source/server/server.cc:558] starting main dispatch loop
[2019-11-06 18:58:29.788][14][info][testing] [test/integration/server.cc:221] stopping integration test server
[2019-11-06 18:58:29.791][38][info][main] [source/server/server.cc:614] shutting down server instance
[2019-11-06 18:58:29.792][38][info][main] [source/server/server.cc:564] main dispatch loop exited
external/bazel_tools/tools/test/test-setup.sh: line 310:    14 Segmentation fault      "${TEST_PATH}" "$@" 2>&1

Reply:
Sorry for the delay @antoniovicente I will look at this PR this week.

Reply:
Yeah you will have to serialize this correctly I think. As long as the dispatcher is still valid memory I think it should work, though there is no guarantee that the event will be run. Can you provide more info on the crashing call stack?